usdigitalresponse/grants-ingest
# 260 Add CLI tool for uploading FFIS backlog to S3

### Ticket #260 

Closes #260 

## Description

This PR adds a CLI-based companion utility for the grants-ingest service. Initially, this tool provides a single `ffis-import`  subcommand that processes a local directory containing FFIS spreadsheets (e.g. backlogs) and uploads them to S3 in order of their published date (oldest first).

Additionally, a few fixes are included based on things I noticed during testing:
1. Certain WARN-level logs were formatted incorrectly due to a missing field/key.
2. `DueDate` parsing logic was updated to attempt two different date formats (presumably which vary because of Excel).
3. Adjusted logging behavior in `PersistFFISData` to be slightly more informative.
4. Modified `PersistFFISData` and `PersistGrantsGovXMLDB` to log a warning rather than failing when no update is made to DynamoDB (because the extant item matches the data in the update operation), which manifests as a `ConditionalCheckFailedException` error.

Additional information:
- Uses https://github.com/alecthomas/kong for CLI implementation
- Shell completion is available for those who love/need/appreciate it
- The binary can be compiled by running `task build-cli`. Builds output to `./bin/grants-ingest` by default.

## Testing

A "dry-run" mode is supported for testing without actually uploading to S3 (and does not require an environment AWS/LocalStack credentials):
```shell
$ ./bin/grants-ingest ffis-import /path/to/xlsx/files fake-bucket --dry-run
```

For more thorough testing:
1. Create/choose an existing S3 (or LocalStack) S3 bucket. For completeness, you can target the grants-ingest "source data" bucket deployed to a LocalStack environment, which will kick off the `SplitFFISSpreadsheet` Lambda as files are uploaded.
5. Create a directory containing _only_ FFIS spreadsheets (you can use the directory at `cmd/SplitFFISSpreadsheet/fixtures` for this, although may want to change the published date (at cell N2) of at least one of the files to be a unique date).
6. Run a command similar to the following:
    ```shell
    $ ./bin/grants-ingest ffis-import /path/to/xlsx/files source-data-bucket --log-level debug
    ```
7. Verify that the files were uploaded to the correct keys (`sources/YYYY/MM/DD/ffis.org/download.xlsx`, where the date components match the published date provided by the respective spreadsheet) in the specified bucket.

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
